src/VocabularyReferenceFieldsManager.php

```diff
@@ -0,0 +1,250 @@
+   *
+   * @param \Drupal\Core\Field\FieldDefinitionInterface $reference_field
+   *   The open vocabulary reference field.
+   * @param \Drupal\Core\Entity\EntityTypeInterface $entity_type
+   *   The entity type definition.
+   * @param string $bundle
+   *   The bundle we are generating fields for.
+   *
+   * @return \Drupal\Core\Field\BaseFieldDefinition[]
+   *   An array of base field definitions.
+   */
+  protected function generateEntityReferenceFields(FieldDefinitionInterface $reference_field, EntityTypeInterface $entity_type, string $bundle): array {
+    $vocabulary_storage = $this->entityTypeManager->getStorage('open_vocabulary');
+    $fields = [];
+
+    $associations = $this->loadAssociationsByField($reference_field->id());
+    foreach ($associations as $association) {
+      /** @var \Drupal\open_vocabularies\OpenVocabularyInterface $vocabulary */
+      $vocabulary = $vocabulary_storage->load($association->getVocabulary());
+      $plugin = $this->referenceHandlerManager->getDefinition($vocabulary->getHandler());
+
+      // Generate an unique name for the field. Prevent long names by generating
+      // a hashed and truncated suffix.
+      $field_name = $association->getName() . '_' . substr(hash('sha256', $reference_field->getName()), 0, 10);
+      $fields[$field_name] = BaseFieldDefinition::create('entity_reference')
+        ->setLabel($association->label())
+        ->setDescription($association->getHelpText())
+        ->setTranslatable(FALSE)
+        ->setRequired($association->isRequired())
+        ->setCardinality($association->getCardinality())
+        ->setSetting('target_type', $plugin['target_type'])
+        ->setSetting('handler_settings', $vocabulary->getHandlerSettings())
+        ->setSetting('open_vocabulary_association', $association->id())
+        ->setSetting('open_vocabulary_reference_field', $reference_field->getName())
+        ->setComputed(TRUE)
+        ->setClass(ComputedVocabularyReferenceFieldItemList::class)
+        ->setReadOnly(FALSE)
+        // Hide the field by default. It will be placed based on its reference
+        // field weight during hook_entity_form_display_alter().
+        // @todo Allow users to choose if an association will have its fields
+        //   configurable.
+        ->setDisplayConfigurable('form', FALSE)
+        // Mimic the data added to fields by entity_bundle_field_info().
+        ->setProvider('open_vocabularies')
+        ->setName($field_name)
+        ->setTargetEntityTypeId($entity_type->id())
+        ->setTargetBundle($bundle);
+    }
+
+    return $fields;
+  }
+
+  /**
+   * Loads association entities that are related to a specific field.
+   *
+   * @param string $field_id
+   *   The field ID.
+   *
+   * @return \Drupal\open_vocabularies\OpenVocabularyAssociationInterface[]
+   *   The loaded associations, ordered by weight.
+   */
+  protected function loadAssociationsByField(string $field_id): array {
```

Comment:
Let's move this to the `OpenVocabularyAssociationStorage` no?

Reply:
I am lazy <3